### PR TITLE
updates reference level, includes all fitted models in objects

### PR DESCRIPTION
**Pull request by sinanpl**
This PR introduces more accurate reference levels and extra output to improve debugging

Breaking changes: 
- group_var should be factor from now on where reference level is the second level of the factor variable. the first level will be introduced in the model as a dummy when pooled model argument specifies the Jann(2008) model.
- output object contains 4 fitted models (model for A, model for B and pooled models (jann / neumark)

``` r
formula = real_wage ~ age | gender
data = OaxacaBlinder::chicago_long

# oacaxa works with binary / logical
datasetv2 = data
datasetv2$gender = ifelse(datasetv2$gender == "male", 1, 0)

modv1_twofold_neumark = OaxacaBlinder::OaxacaBlinderDecomp(
  formula=formula,
  data = data,
  type = 'twofold',
  pooled = 'neumark',
  baseline_invariant = FALSE,
  n_bootstraps = NULL
)
modv1_twofold_jann = OaxacaBlinder::OaxacaBlinderDecomp(
  formula=formula,
  data = data,
  type = 'twofold',
  pooled = 'jann',
  baseline_invariant = FALSE,
  n_bootstraps = NULL
)
modv2 = oaxaca::oaxaca(formula = formula, data = datasetv2, R=NULL)
#> oaxaca: oaxaca() performing analysis. Please wait.

# check outputs oaxaca for neumark
summary(modv1_twofold_neumark)
#> Oaxaca Blinder Decomposition model
#> ----------------------------------
#> Type: twofold
#> Formula: real_wage ~ age | gender
#> Data: data
#> 
#> Descriptives
#>                  n    %n mean(real_wage)
#> gender==male   412 57.9%           13.69
#> gender==female 300 42.1%           17.52
#> 
#> Gap: -3.83
#> % Diff: -28.01%
#>               coefficient   % of gap
#> explained           -0.20       5.3%
#> unexplained         -3.63      94.7%
#> unexplained_a       -2.06      53.7%
#> unexplained_b       -1.57      40.9%
t(t(modv2$twofold$overall[c(5), ]))
#>                          [,1]
#> group.weight        -1.000000
#> coef(explained)     -0.204005
#> se(explained)              NA
#> coef(unexplained)   -3.630331
#> se(unexplained)            NA
#> coef(unexplained A) -2.060458
#> se(unexplained A)          NA
#> coef(unexplained B) -1.569873
#> se(unexplained B)          NA

# check outputs oaxaca for jann
summary(modv1_twofold_jann)
#> Oaxaca Blinder Decomposition model
#> ----------------------------------
#> Type: twofold
#> Formula: real_wage ~ age | gender
#> Data: data
#> 
#> Descriptives
#>                  n    %n mean(real_wage)
#> gender==male   412 57.9%           13.69
#> gender==female 300 42.1%           17.52
#> 
#> Gap: -3.83
#> % Diff: -28.01%
#>               coefficient   % of gap
#> explained           -0.19       5.0%
#> unexplained         -3.64      95.0%
#> unexplained_a        0.00       0.0%
#> unexplained_b       -3.64      95.0%
t(t(modv2$twofold$overall[c(6), ]))
#>                              [,1]
#> group.weight        -2.000000e+00
#> coef(explained)     -1.918078e-01
#> se(explained)                  NA
#> coef(unexplained)   -3.642528e+00
#> se(unexplained)                NA
#> coef(unexplained A) -1.940510e-14
#> se(unexplained A)              NA
#> coef(unexplained B) -3.642528e+00
#> se(unexplained B)              NA

# check existence of all models in output object
str(modv1_twofold_jann$meta$fitted_models, max.level = 1)
#> List of 4
#>  $ mod_a                 :List of 13
#>   ..- attr(*, "class")= chr "lm"
#>  $ mod_b                 :List of 13
#>   ..- attr(*, "class")= chr "lm"
#>  $ mod_pooled_neumark1988:List of 13
#>   ..- attr(*, "class")= chr "lm"
#>  $ mod_pooled_jann2008   :List of 14
#>   ..- attr(*, "class")= chr "lm"
```

<sup>Created on 2024-03-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by sinanpl** [comment]
@davidskalinder what do you think of this? 

this is an alternative + some extras for https://github.com/sinanpl/OaxacaBlinder/pull/11

File: R/oaxaca.R [hunk 1]
```diff
@@ -93,7 +86,7 @@ fit_models <- function(formula, data) {
   mod_a = lm(fml_reg, data = data_a)
   mod_b = lm(fml_reg, data = data_b)
   mod_pooled_neumark1988 = lm(fml_reg_pooled_neumark1988, data = data)
-  mod_pooled_jann2008 = lm(fml_reg_pooled_jann2008, data = data)
+  mod_pooled_jann2008 = lm(fml_reg_pooled_jann2008, data = data, contrasts = )
```
Comment (davidskalinder) [inline R/oaxaca.R:89]: So I don't think this breaks anything, but obviously that trailing `contrasts = ` arg shouldn't be there.  I confess I am bad at thinking through contrasts matrices so I can't tell whether line 61 above handles this properly -- @sinanpl I guess have a close look at this to make sure it's doing what you want with the contrast matrix for the Jann version?

**Review by davidskalinder — CHANGES_REQUESTED** [review]
This looks pretty clean to me except for the business I noted in the code with the contrast matrices, as well as the bad call to the new `parse_formula()` that should be fixed in #19.  @sinanpl, doing this made me realize that I've never reviewed a PR before, so let me know if I'm screwing anything up here!  Hopefully GitHub has made the process un-screw-uppable.

As far as the design is concerned, like I said in my comment at https://github.com/sinanpl/OaxacaBlinder/pull/11#issuecomment-2015494016, if I had my druthers I'd prefer an option to make the baseline choice explicit rather than relying on the factor structure (plus my suggestion there would preserve the handy ability to force a positive or negative gap!).  But this version does the most important thing, which is to allow the user to control the baseline somehow; so @sinanpl if you prefer this version then I'm happy to go with it.

Two other slight concerns I'd have about this PR:

1. The new behavior should get documentation ASAP since otherwise it's pretty tricky to discover what's going on.
2. It'd be nice to have some tests for the new behavior, but as I mentioned in https://github.com/sinanpl/OaxacaBlinder/issues/7#issuecomment-2010711683 I don't think the tests have been merged in yet?  I think we could use the [reprex you posted](https://github.com/sinanpl/OaxacaBlinder/pull/18#issue-2204036056) for some tests, and it seems like it'd be good to test that the factor variable can be flipped correctly?

File: R/oaxaca.R (hunk 1 above)
Comment (sinanpl) [reply R/oaxaca.R]: yeah, i just did some tests with the flipped versions, did not work as expected. needs more time to investigate. 
i did mrerge the initial tests branch (with the flipped level tests) into this one to make development more convenient, FYI

Comment (sinanpl) [reply R/oaxaca.R]: the contrasts are actually only relevant for the pooled model including the group variable. however, the flipped factor does not produce results that are equivalent to the oaxaca package. the root cause for this is somewhere else I assume, to be investigated

Comment (sinanpl) [reply R/oaxaca.R]: The code in `/R` was actually ok, the cause was in flipping the level in the creation of the flipped level data 🤦‍♂️

``` r
x = factor(c("a", "a", "b"))
y = factor(c("c", "c", "d"))

# bug
levels(x) = rev(levels(x))
x
#> [1] b b a
#> Levels: b a

# correct relevel
y = relevel(y, ref = levels(y)[2])
y
#> [1] c c d
#> Levels: d c
```

<sup>Created on 2024-03-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/oaxaca.R [hunk 2]
```diff
@@ -304,9 +304,24 @@ get_bootstrap_ci = function(formula,
 #' @return A list with elements \code{overall}, \code{varlevel}, \code{gaps},
 #'   \code{meta}, and \code{bootstraps}, which can be queried with
 #'   \code{summary()} and \code{coef()}.
-#' @export
+#' @details
+#' \subsection{Reference level of grouping variable}{
+#'  In this package, the main function requires the grouping variable
+#'  in the data to be a factor with 2 levels. The statistical method
+#'  will calculate the gaps based on the 1st level of the factor.
```
Comment (davidskalinder) [inline R/oaxaca.R:311]: This isn't 100% clear to me yet -- does this mean "the gaps will be calculated by subtracting the mean outcome for the first level of the factor from the mean outcome for the second level (i.e., \eqn{\Delta \bar{Y} = \bar{Y}_2 - \bar{Y}_1)"? (NB the latex is air code.)

File: R/oaxaca.R [hunk 3]
```diff
@@ -304,9 +304,24 @@ get_bootstrap_ci = function(formula,
...
+#'
+#'  Also, in case the pooled model should contain the grouping
+#'  variable, the first factor level is the level that is encoded to a dummy.
```
Comment (davidskalinder) [inline R/oaxaca.R:314]: Likewise here, I think I'm getting turned around -- this makes it sound like a variable `sex` with first level `m` and second level `f` will be treated as a dummy for `male`.  Is that correct?

Whichever way it is, it seems to me like it'd be clearer to specify which level is used as the baseline/reference category?

**Comment by davidskalinder** [comment]
Some notes here that I don't know what to do with yet:

In `oaxaca`, observations for which the group variable takes the value 1/`TRUE` get assigned to group B, whereas 0/`FALSE` gets assigned to group A:

```
> oaxaca::oaxaca(ln.real.wage ~ foreign.born | female, oaxaca::chicago, R = NULL) |> purrr::pluck("y")
oaxaca: oaxaca() performing analysis. Please wait.
$y.A
[1] 2.704193

$y.B
[1] 2.498258

$y.diff
[1] 0.2059347

> oaxaca::chicago |> dplyr::group_by(female) |> dplyr::summarize(mean = mean(ln.real.wage, na.rm = TRUE)) |> as.data.frame()
  female     mean
1      0 2.704193
2      1 2.498258
```

After loading the branch in this PR, observations for which the group variable takes a value of the second factor level get assigned to group A, whereas the first factor level gets assigned to group B:

```
> OaxacaBlinderDecomp(ln_real_wage ~ foreign_born | gender, chicago_long) |> purrr::pluck("gaps")
$gap
[1] -0.2059347

$pct_gap
[1] -0.08243132

$EY_a
[1] 2.498258

$EY_b
[1] 2.704193

> chicago_long |> dplyr::group_by(gender) |> dplyr::summarize(mean = mean(ln_real_wage, na.rm = TRUE)) |> as.data.frame()
  gender     mean
1   male 2.704193
2 female 2.498258

> levels(chicago_long$gender)
[1] "male"   "female"
```

... as expected given the assignments in `parse_formula()` (line 13, currently) and `fit_models()` (lines 63-64, currently).

So to summarize:

package | group_A_val | group_B_val
-- | -- | --
oaxaca | 0 or FALSE | 1 or TRUE
OaxacaBlinder | second level | first level

It seems to me that the implementation above is therefore the wrong way around?

NB that I think part of the problem might be that the language is a bit fuzzy.  Jann (2008:455) doesn't use the term "reference" for a value of the group variable, but instead notes that when the decomp is performed in the usual way (subtracting group B from group A, same as in `calculate_coefs`), the decomp is "formulated from the viewpoint of group B".  I think we should probably use that language to avoid getting turned around.

I should also check how Stata defines the groups by default and add it to that table.  Might not get to that this evening though.

**Comment by davidskalinder** [comment]
Here's how Jann's Stata command handles it:

```
. oaxaca ln_real_wage age normalize(LTHS some_college college high_school advanced_degree), by(female)
(normalized: LTHS some_college college high_school advanced_degree)

Blinder-Oaxaca decomposition                    Number of obs     =        666
                                                Model             =     linear
Group 1: female = 0                             N of obs 1        =        378
Group 2: female = 1                             N of obs 2        =        288

   endowments: (X1 - X2) * b2
 coefficients: X2 * (b1 - b2)
  interaction: (X1 - X2) * (b1 - b2)
```

So the full table is like so:

package | group_A_val | group_B_val
-- | -- | --
oaxaca (Hlavac) | 0 or FALSE | 1 or TRUE
oaxaca (Jann) | 0 | 1
OaxacaBlinder | second level | first level

Of course, the levels of a factor aren't really the same as the two values of a logical, but I think it's natural to view the `FALSE` value as being like the first level and the `TRUE` value as the second; so if we want to stick to the implementation of the other packages, the direction of the implementation in this PR should be flipped.

Having said that: I'm not entirely convinced that that's a bad thing, for several reasons:

1. It seems counterintuitive to me to subtract the `TRUE` group variable from the `FALSE` one.  (Though I suppose it is more intuitive if the `TRUE` group is often a disempowered group whose disempowerment is being investigated?  So, for instance, it makes sense to do the decomposition from the viewpoint of `female == TRUE`?)
2. `lm()` (and probably other models in R?) by default uses the first level of a factor as the reference level.  (It's not entirely clear to me how Stata chooses the reference level.)
3. `relevel()` only sets the first level of a factor (which it calls the "reference level"); so it can easily be used to set the first level, but not the second.

I think item 3 is the most serious of these problems, since users need a way to set the "viewpoint" level easily.  If they can't do that using `relevel()`, then I think we should export another function to do it instead (which can probably just be a function that flips `relevel()` but is more intuitively named).

@sinanpl, any thoughts about this?

**Comment by davidskalinder** [comment]
I'll add BTW that personally I still feel like the best way to handle this is just to have `OaxacaBlinderDecomp()` take a new option to specify the value of the viewpoint group.  If we follow through on this PR's proposal to drop the auto-group-choice behavior, then this doesn't even have the problem of requiring two options as discussed in https://github.com/sinanpl/OaxacaBlinder/pull/11#issuecomment-2015494016.  I think allowing users to directly specify the viewpoint level is clearer and more straightforward, and of course we can still set the default group how we want depending on the considerations above...

**Comment by davidskalinder** [comment]
So @sinanpl my plan was to try to complete this PR and get it into the codebase so I could start using (and testing!) the package in my project.  However, thinking about it, I don't want to make the call on how to handle what I guess I'll now (until I find a better name?) start calling the "viewpoint" group until you weigh in (i.e., on whether to use the first factor level as the other packages do, to use the second level as this PR does, or to have a user-specified option with one of the levels as the default).  Obviously any changes to that setup would be breaking changes, and possibly confusing ones at that, so I'm hesitant to send the thing in a direction that's hard to get back from later.

For the time being, I'll implement a new private mod with a simple version of my preferred solution of a user-specified option and use that for testing.
